src/libmeasurement_kit/ooni/facebook_messenger.cpp

```diff
@@ -0,0 +1,228 @@
+                              Var<Logger>> cb
+                    ) {
+    if (error) {
+        cb(error, entry, fb_service_ips, reactor, logger);
+        return;
+    }
+
+    // if we find any inconsistent DNS later, switch this to true
+    (*entry)["facebook_dns_blocking"] = false;
+
+    int names_count = fb_service_hostnames.size();
+    Var<int> names_tested(new int(0));
+
+    if (names_count == *names_tested) {
+        cb(NoError(), entry, fb_service_ips, reactor, logger);
+        return;
+    }
+
+    // XXX pretty this up
+    // maybe just make one of these on the heap in the caller and pass a
+    // reference to this function? instead of making a heap copy here.
+    Var<std::map<std::string, std::vector<std::string>>>
+        fb_service_ips2(new std::map<std::string, std::vector<std::string>>(fb_service_ips));
+
+    auto dns_cb = [=](std::string service, std::string hostname) {
+        return [=](Error err, Var<dns::Message> message) {
+            if (!!err) {
+                logger->info("fb_messenger: dns error for %s, %s",
+                    service.c_str(), hostname.c_str());
+            } else {
+                for (auto answer : message->answers) {
+                    if ((answer.ipv4 != "") || (answer.hostname != "")) {
+                        //XXX
+                        //std::string asn_p = options.get("geoip_asn_path", std::string{});
+                        std::string asn_p = "GeoIPASNum.dat";
+                        GeoipDatabase geoip(asn_p);
+                        ErrorOr<std::string> asn = geoip.resolve_asn(answer.ipv4);
+                        if (asn && asn.as_value() != "AS0") {
+                            logger->info("%s ipv4: %s, %s",
+                                hostname.c_str(), answer.ipv4.c_str(),
+                                asn.as_value().c_str());
+                            // if consistent, add to list for tcp connect later
+                            if (asn.as_value() == "AS32934") {
+                                (*fb_service_ips2)[service].push_back(answer.ipv4);
+                            } else {
+                                (*entry)["facebook_dns_blocking"] = true;
+                            }
+                        }
+                    }
+                }
+                // if any consistent IPs, consistent = true
+                (*entry)["facebook_" + service + "_dns_consistent"] =
+                    !(*fb_service_ips2)[service].empty();
+            }
+            *names_tested += 1;
+            if (names_count == *names_tested) {
+                cb(NoError(), entry, *fb_service_ips2, reactor, logger);
+                return;
+            }
+        };
+    };
+
+    for (auto const& service_and_hostname : fb_service_hostnames) {
+        std::string service = service_and_hostname.first;
+        std::string hostname = service_and_hostname.second;
+        Settings options; // XXX ought this to forward from above?
```

Comment:
Yes, generally we want to bring around the same `settings` object.

Reply:
done

Reply:
done